rfc/20241206-oci-registry-auth-config.md

```diff
@@ -0,0 +1,174 @@
+
+This proposal currently assumes that it's sufficient to support one credentials method per distinct registry domain name, across all operations. That seems to follow the precedent set by other tools in the ecosystem, such as `docker login` being designed to issue one set of credentials for each distinct server domain name.
+
+If we later add features to OpenTofu that involve _writing_ to OCI registries, is it likely that someone would want to use different credentials for the read-only use-cases of package installation than for use-cases that write to the registry? If so, does any other tooling in the ecosystem have some design precedent we can follow in supporting that?
+
+### Future Considerations
+
+This proposal was strongly motivated by the specific use-cases of fetching provider packages and module packages during the dependency installation step of `tofu init`.
+
+This proposal _aims_ to be a cross-cutting piece of infrastructure that we could use for other features that behave as an OCI Distribution client in future. Some that we've already discussed in other locations are:
+
+- Using an OCI repository to store OpenTofu state snapshots ([#1230](https://github.com/opentofu/opentofu/issues/1230), [#1363](https://github.com/opentofu/opentofu/issues/1363)): If we decide to integrate this into OpenTofu CLI, rather than [Backends as Plugins](https://github.com/opentofu/opentofu/issues/382), then it could benefit from the same credentials configuration.
+
+    (On the other hand, if we decide to require state storage to always be in plugins then a plugin for OCI repositories would be a separate program that would need to solve credentials-gathering for itself some other way.)
+- Commands for "pushing" modules and/or providers to a registry (also discussed in [#1672](https://github.com/opentofu/opentofu/issues/1672)): To reduce scope we're likely to focus only on read-only installation use-cases for a first round, but in future we might also offer features for copying module packages from local disk to remote locations, potentially including support for writing to an OCI repository, in which case the OCI-interacting parts of those features should benefit from the same credentials configuration.
+
+## Potential Alternatives
+
+### Support the Docker CLI / Podman CLI configuration files exclusively
+
+Based on an informal survey, it seems relatively common to treat the Docker CLI and Podman CLI configuration files -- which both use a compatible file format but different filesystem search paths -- as the primary location for OCI registry credentials in other tools.
+
+We could potentially decide to follow that lead and not offer any OpenTofu-specific CLI configuration options at all. That would eliminate some features that we need to develop, test, and maintain, but would make it harder for someone to use a separate configuration for OpenTofu than for other tools that interact with OCI registries.
```

Comment:
Does this not also require installation of one of these external tools in addition to OpenTofu if one wants to authenticate with an authenticated registry as a drawback (or hand author the relevant credentials files that Docker/Podman _would_ have generated)?

When running OpenTofu _within_ a container, is the expectation that a user will use Docker in Docker or the like to expose the underlying docker/podman CLI from within the container? Would this necessarily mean that there can't be isolation of authentication profiles between OpenTofu containers, as they all use the same underlying credential helper? 

What if container A has different access to modules hosted by the example.com registry than container B? e.g. container A has access to AWS modules hosted by example.com, but container B has access to Azure modules hosted by example.com. Is this functionality that isn't supported?

Reply:
The general point I'm trying to bring up is that (my guess is that) a user using OpenTofu in the context of a container would likely prefer to use some OpenTofu specific authentication mechanism for accessing these OCI artifacts rather than reusing Docker/Podman authentication mechanisms because of the complications that can arise from either hand-authoring Docker/Podman style configurations or finding a way for the container to have access to the configurations from the host running the container.

I don't actually know how complicated it is to setup this configuration by hand. Do you think the average user would be OK with hand authoring the configurations?

Reply:
Ahh okay, I understand!

Note that this text is under the "Potential Alternatives" and so this is describing a potential option for changing the proposal but it is not actually part of the proposal.

At the time I'm writing this the proposal is to have a separate configuration mechanism specific to open OpenTofu and to use the Docker CLI / Podman CLI configuration sources only as a fallback when no OpenTofu-specific settings are present.

Our RFC format asks authors to describe potential alternatives and so I wrote this here after my research suggested that some other tools in the OCI ecosystem just rely directly on Docker CLI / Podman CLI configuration conventions, but I'm personally in favor of offering an OpenTofu-specific mechanism. My reason for that is perhaps a little different than yours -- I expect that folks configuring OpenTofu execution environments would benefit from having only a single CLI configuration file to manage in that environment -- but I think we have arrived at the same conclusion anyway.